python/taichi/main.py

```diff
@@ -1,358 +1,481 @@
-        pytest_args = [test_dir]
-    if args.verbose:
-        pytest_args += ['-s', '-v']
-    if args.rerun:
-        pytest_args += ['--reruns', args.rerun]
-    if int(
-            pytest.main(
-                [os.path.join(root_dir, 'misc/empty_pytest.py'), '-n1',
-                 '-q'])) == 0:  # test if pytest has xdist or not
-        try:
-            from multiprocessing import cpu_count
-            threads = min(8, cpu_count())  # To prevent running out of memory
-        except:
-            threads = 2
-        os.environ['TI_DEVICE_MEMORY_GB'] = '0.5'  # Discussion: #769
-        arg_threads = None
-        if args.threads is not None:
-            arg_threads = int(args.threads)
-        env_threads = os.environ.get('TI_TEST_THREADS', '')
-        if arg_threads is not None:
-            threads = arg_threads
-        elif env_threads:
-            threads = int(env_threads)
-        print(f'Starting {threads} testing thread(s)...')
-        if threads > 1:
-            pytest_args += ['-n', str(threads)]
-    return int(pytest.main(pytest_args))
-
-
-def test_cpp(args):
-    import taichi as ti
-    test_files = args.files
-    # Cpp tests use the legacy non LLVM backend
-    ti.reset()
-    print("Running C++ tests...")
-    task = ti.Task('test')
-    return int(task.run(*test_files))
-
-
-def get_examples_dir() -> Path:
-    """Get the path to the examples directory."""
-    import taichi as ti
-
-    root_dir = ti.package_root() if ti.is_release() else ti.get_repo_directory(
-    )
-    examples_dir = Path(root_dir) / 'examples'
-    return examples_dir
-
-
-def get_available_examples() -> set:
-    """Get a set of all available example names."""
-    examples_dir = get_examples_dir()
-    all_examples = examples_dir.rglob('*.py')
-    all_example_names = {
-        str(f.resolve()).split('/')[-1].split('.')[0]
-        for f in all_examples
+from colorama import Back, Fore, Style
+
+import taichi as ti
+from taichi.tools.video import (accelerate_video, crop_video, make_video, mp4_to_gif, scale_video)
+
+
+def timer(func):
+    import timeit
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        start = timeit.default_timer()
+        result = func(*args, **kwargs)
+        elapsed = timeit.default_timer() - start
+        print(f">>> Running time: {elapsed:.2f}s")
+        return result
+    return wrapper
+
+
+class TaichiMain:
+    registered_commands = {
+        "example",
+        "release",
+        "gif",
+        "video_speed",
+        "video_crop",
+        "video_scale",
+        "video",
+        "doc",
+        "update",
+        "format_all",
+        "format",
+        "build",
+        "regression",
+        "baseline",
+        "benchmark",
+        "test",
+        "debug",
+        "run"
     }
-    return all_example_names
 
+    @timer
+    def __init__(self, debug=False):
+        self.banner = f"\n{'*' * 43}\n**   \u262f Taichi Programming Language       **\n{'*' * 43}"
```

Comment:
> UnicodeEncodeError: 'charmap' codec can't encode character '\u262f' in position 52: character maps to <undefined>
Command exited with code 1

Windows console does not like this unicode, will need to change the encoding to UTF-8 or remove it..

Reply:
Let's remove then. Portability is more important than a fancy logo :-)

Reply:
Of course ;)